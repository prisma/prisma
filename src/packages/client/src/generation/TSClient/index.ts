import 'flat-map-polyfill' // unfortunately needed as it's not properly polyfilled in TypeScript

export { MinimalArgsType, ArgsType } from './Args'
export { Enum } from './Enum'
export { JS, TS } from './Generatable'
export { InputField, InputType } from './Input'
export { Model, ModelDelegate } from './Model'
export { OutputField, OutputType } from './Output'
export { SchemaOutputField, SchemaOutputType } from './SchemaOutput'
export { TSClient } from './TSClient'

