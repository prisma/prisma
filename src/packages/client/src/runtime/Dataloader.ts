interface Job {
  resolve: (data: any) => void
  reject: (data: any) => void
  request: any
}

export type DataloaderOptions<T> = {
  singleLoader: (request: T) => Promise<any>
  batchLoader: (request: T[]) => Promise<any[]>
  batchBy: (request: T) => string | null
}

export class Dataloader<T = any> {
  batches: { [key: string]: Job[] }
  private tickActive: boolean = false
  constructor(private options: DataloaderOptions<T>) {
    this.batches = {}
  }
  request(request: T): Promise<any> {
    const hash = this.options.batchBy(request)
    if (!hash) {
      return this.options.singleLoader(request)
    }
    if (!this.batches[hash]) {
      this.batches[hash] = []

      // make sure, that we only tick once at a time
      if (!this.tickActive) {
        this.tickActive = true
        process.nextTick(() => {
          this.dispatchBatches()
          this.tickActive = false
        })
      }
    }

    return new Promise((resolve, reject) => {
      this.batches[hash].push({
        request,
        resolve,
        reject,
      })
    })
  }

  private dispatchBatches() {
    for (const key in this.batches) {
      const batch = this.batches[key]
      delete this.batches[key]

      // only batch if necessary
      // this might occur, if there's e.g. only 1 findUnique in the batch
      if (batch.length === 1) {
        this.options
          .singleLoader(batch[0].request)
          .then((result) => {
            if (result instanceof Error) {
              batch[0].reject(result)
            } else {
              batch[0].resolve(result)
            }
          })
          .catch((e) => {
            batch[0].reject(e)
          })
      } else {
        this.options
          .batchLoader(batch.map((j) => j.request))
          .then((results) => {
            if (results instanceof Error) {
              for (let i = 0; i < batch!.length; i++) {
                batch![i].reject(results)
              }
            } else {
              for (let i = 0; i < batch!.length; i++) {
                const value = results[i]
                if (value instanceof Error) {
                  batch![i].reject(value)
                } else {
                  batch![i].resolve(value)
                }
              }
            }
          })
          .catch((e) => {
            for (let i = 0; i < batch!.length; i++) {
              batch![i].reject(e)
            }
          })
      }
    }
  }
}
