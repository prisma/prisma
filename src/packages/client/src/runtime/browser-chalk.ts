const identity = (str) => str || ''

const browserChalk: any = new Proxy(identity, {
  get: (obj, prop) => {
    return browserChalk
  },
  apply: (target, that, args) => {
    return target(args[0])
  },
})

export default browserChalk
