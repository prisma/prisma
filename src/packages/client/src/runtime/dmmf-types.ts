export namespace DMMF {
  export interface Document {
    datamodel: Datamodel
    schema: Schema
    mappings: Mapping[]
  }

  export interface Enum {
    name: string
    values: string[]
    dbName?: string | null
    documentation?: string
  }

  export interface Datamodel {
    models: Model[]
    enums: Enum[]
  }

  export interface uniqueIndex {
    name: string
    fields: string[]
  }

  export interface Model {
    name: string
    isEmbedded: boolean
    dbName: string | null
    fields: Field[]
    fieldMap?: Record<string, Field>
    uniqueFields: string[][]
    uniqueIndexes: uniqueIndex[]
    documentation?: string
    idFields: string[]
    [key: string]: any // safe net for additional new props
  }

  export type FieldKind = 'scalar' | 'object' | 'enum'

  export interface Field {
    kind: FieldKind
    name: string
    isRequired: boolean
    isList: boolean
    isUnique: boolean
    isId: boolean
    type: string
    dbNames: string[] | null
    isGenerated: boolean
    hasDefaultValue: boolean
    default?: FieldDefault | string | boolean | number
    relationToFields?: any[]
    relationOnDelete?: string
    relationName?: string
    documentation?: string
    [key: string]: any // safe net for additional new props
  }

  export interface FieldDefault {
    name: string
    args: any[]
  }

  export interface Schema {
    rootQueryType?: string
    rootMutationType?: string
    inputTypes: InputType[]
    outputTypes: OutputType[]
    enums: Enum[]
  }

  export interface Query {
    name: string
    args: SchemaArg[]
    output: QueryOutput
  }

  export interface QueryOutput {
    name: string
    isRequired: boolean
    isList: boolean
  }

  export type ArgType = string | InputType | Enum

  export interface SchemaArgInputType {
    isRequired: boolean
    isNullable: boolean
    isList: boolean
    type: ArgType
    kind: FieldKind
  }

  export interface SchemaArg {
    name: string
    inputType: SchemaArgInputType[]
    isRelationFilter?: boolean
    nullEqualsUndefined?: boolean
    comment?: string
  }

  export interface OutputType {
    name: string
    fields: SchemaField[]
    fieldMap?: Record<string, SchemaField>
    isEmbedded?: boolean
  }

  export interface SchemaField {
    name: string
    outputType: {
      type: string | OutputType | Enum // note that in the serialized state we don't have the reference to MergedOutputTypes
      isList: boolean
      isRequired: boolean
      kind: FieldKind
    }
    args: SchemaArg[]
  }

  export interface InputType {
    name: string
    isWhereType?: boolean // this is needed to transform it back
    isOrderType?: boolean
    isUpdateType?: boolean
    isUpdateOperationType?: boolean
    atLeastOne?: boolean
    atMostOne?: boolean
    isOneOf?: boolean
    fields: SchemaArg[]
    fieldMap?: Record<string, SchemaArg>
  }

  export interface Mapping {
    model: string
    plural: string
    findOne?: string | null
    findMany?: string | null
    create?: string | null
    update?: string | null
    updateMany?: string | null
    upsert?: string | null
    delete?: string | null
    deleteMany?: string | null
    aggregate?: string | null
  }

  export enum ModelAction {
    findOne = 'findOne',
    findMany = 'findMany',
    create = 'create',
    update = 'update',
    updateMany = 'updateMany',
    upsert = 'upsert',
    delete = 'delete',
    deleteMany = 'deleteMany',
  }
}

export interface BaseField {
  name: string
  type: string | DMMF.Enum | DMMF.OutputType | DMMF.SchemaArg
  isList: boolean
  isRequired: boolean
}
