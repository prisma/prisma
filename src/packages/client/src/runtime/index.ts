export { DMMF } from './dmmf-types'
export { DMMFClass } from './dmmf'
export {
  makeDocument,
  transformDocument,
  unpack,
  PrismaClientValidationError,
} from './query'
export { default as debugLib } from '@prisma/debug'

export {
  Engine,
  PrismaClientKnownRequestError,
  PrismaClientUnknownRequestError,
  PrismaClientInitializationError,
  PrismaClientRustPanicError,
} from '@prisma/engine-core'
export { getPrismaClient, PrismaClientOptions } from './getPrismaClient'

export {
  RawValue,
  Sql,
  Value,
  empty,
  join,
  raw,
  sqltag,
} from 'sql-template-tag'


export { default as Decimal } from 'decimal.js'
