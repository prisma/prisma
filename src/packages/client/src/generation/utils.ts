import indent from 'indent-string'
import path from 'path'
import { DMMFClass } from '../runtime/dmmf'
import { DMMF } from '../runtime/dmmf-types'

export enum Projection {
  select = 'select',
  include = 'include',
}

export function getScalarsName(modelName: string): string {
  return `${modelName}Scalars`
}

export function getPayloadName(modelName: string): string {
  return `${modelName}GetPayload`
}

// export function getExtractName(modelName: string, projection: Projection) {
//   return `Extract${modelName}${capitalize(projection)}`
// }

export function getSelectName(modelName: string): string {
  return `${modelName}Select`
}

export function getAggregateName(modelName: string): string {
  return `Aggregate${capitalize(modelName)}`
}

export function getAvgAggregateName(modelName: string): string {
  return `${capitalize(modelName)}AvgAggregateOutputType`
}

export function getSumAggregateName(modelName: string): string {
  return `${capitalize(modelName)}SumAggregateOutputType`
}

export function getMinAggregateName(modelName: string): string {
  return `${capitalize(modelName)}MinAggregateOutputType`
}

export function getMaxAggregateName(modelName: string): string {
  return `${capitalize(modelName)}MaxAggregateOutputType`
}

export function getAggregateInputType(aggregateOutputType: string): string {
  return aggregateOutputType.replace(/OutputType$/, 'InputType')
}

export function getAggregateArgsName(modelName: string): string {
  return `Aggregate${capitalize(modelName)}Args`
}

export function getAggregateGetName(modelName: string): string {
  return `Get${capitalize(modelName)}AggregateType`
}

export function getAggregateScalarGetName(modelName: string): string {
  return `Get${capitalize(modelName)}AggregateScalarType`
}

export function getIncludeName(modelName: string): string {
  return `${modelName}Include`
}

export function getDefaultName(modelName: string): string {
  return `${modelName}Default`
}

export function getFieldArgName(field: DMMF.SchemaField): string {
  return getArgName(
    (field.outputType.type as DMMF.OutputType).name,
    field.outputType.isList,
  )
}

export function getArgName(name: string, isList: boolean): string {
  if (!isList) {
    return `${name}Args`
  }

  return `FindMany${name}Args`
}

// we need names for all top level args,
// as GraphQL doesn't have the concept of unnamed args
export function getModelArgName(
  modelName: string,
  action?: DMMF.ModelAction,
): string {
  if (!action) {
    return `${modelName}Args`
  }
  switch (action) {
    case DMMF.ModelAction.findMany:
      return `FindMany${modelName}Args`
    case DMMF.ModelAction.findOne:
      return `FindOne${modelName}Args`
    case DMMF.ModelAction.findFirst:
      return `FindFirst${modelName}Args`
    case DMMF.ModelAction.upsert:
      return `${modelName}UpsertArgs`
    case DMMF.ModelAction.update:
      return `${modelName}UpdateArgs`
    case DMMF.ModelAction.updateMany:
      return `${modelName}UpdateManyArgs`
    case DMMF.ModelAction.delete:
      return `${modelName}DeleteArgs`
    case DMMF.ModelAction.create:
      return `${modelName}CreateArgs`
    case DMMF.ModelAction.deleteMany:
      return `${modelName}DeleteManyArgs`
  }
}

export function getDefaultArgName(
  dmmf: DMMFClass,
  modelName: string,
  action: DMMF.ModelAction,
): string {
  const mapping = dmmf.mappings.find((m) => m.model === modelName)!

  const fieldName = mapping[action]
  const operation = getOperation(action)
  const queryType = operation === 'query' ? dmmf.queryType : dmmf.mutationType
  const field = queryType.fields.find((f) => f.name === fieldName)!
  return (field.args[0].inputTypes[0].type as DMMF.InputType).name
}

export function getOperation(action: DMMF.ModelAction): 'query' | 'mutation' {
  if (
    action === DMMF.ModelAction.findMany ||
    action === DMMF.ModelAction.findOne
  ) {
    return 'query'
  }
  return 'mutation'
}

/**
 * Used to render the initial client args
 * @param modelName
 * @param fieldName
 * @param mapping
 */
export function renderInitialClientArgs(
  actionName: DMMF.ModelAction,
  fieldName: string,
  mapping: DMMF.Mapping,
): string {
  return `
  dmmf,
  fetcher,
  '${getOperation(actionName)}',
  '${fieldName}',
  '${mapping.plural}.${actionName}',
  args || {},
  [],
  errorFormat,
  measurePerformance\n`
}

export function getFieldTypeName(field: DMMF.SchemaField): string {
  if (typeof field.outputType.type === 'string') {
    return field.outputType.type
  }

  return field.outputType.type.name
}

export function getType(
  name: string,
  isList: boolean,
  isOptional?: boolean,
): string {
  return name + (isList ? '[]' : '') + (isOptional ? ' | null' : '')
}

export function getFieldType(field: DMMF.SchemaField): string {
  return getType(getFieldTypeName(field), field.outputType.isList)
}

interface SelectReturnTypeOptions {
  name: string
  actionName: DMMF.ModelAction
  renderPromise?: boolean
  hideCondition?: boolean
  isField?: boolean
  fieldName?: string
  projection: Projection
}

/**
 * Get the complicated extract output
 * @param name Model name
 * @param actionName action name
 */
export function getSelectReturnType({
  name,
  actionName,
  renderPromise = true,
  hideCondition = false,
  isField = false, // eslint-disable-line @typescript-eslint/no-unused-vars
}: SelectReturnTypeOptions): string {
  const isList = actionName === DMMF.ModelAction.findMany

  if (actionName === 'deleteMany' || actionName === 'updateMany') {
    return `Promise<BatchPayload>`
  }

  /**
   * Important: We handle findMany or isList special, as we don't want chaining from there
   */
  if (isList || hideCondition) {
    const listOpen = isList ? 'Array<' : ''
    const listClose = isList ? '>' : ''
    const promiseOpen = renderPromise ? 'Promise<' : ''
    const promiseClose = renderPromise ? '>' : ''

    return `CheckSelect<T, ${promiseOpen}${listOpen}${name}${listClose}${promiseClose}, ${promiseOpen}${listOpen}${getPayloadName(
      name,
    )}<T>${listClose}${promiseClose}>`
  }

  return `CheckSelect<T, Prisma__${name}Client<${getType(name, isList)}${actionName === 'findOne' ? ' | null' : ''
    }>, Prisma__${name}Client<${getType(getPayloadName(name) + '<T>', isList)}${actionName === 'findOne' ? ' | null' : ''
    }>>`
}

export function isQueryAction(
  action: DMMF.ModelAction,
  operation: 'query' | 'mutation',
): boolean {
  if (!(action in DMMF.ModelAction)) {
    return false
  }
  const result =
    action === DMMF.ModelAction.findOne || action === DMMF.ModelAction.findMany
  return operation === 'query' ? result : !result
}

export function capitalize(str: string): string {
  return str[0].toUpperCase() + str.slice(1)
}

export function indentAllButFirstLine(
  str: string,
  indentation: number,
): string {
  const lines = str.split('\n')

  return lines[0] + '\n' + indent(lines.slice(1).join('\n'), indentation)
}

export function getRelativePathResolveStatement(
  outputDir: string,
  cwd?: string,
): string {
  if (!cwd) {
    return 'undefined'
  }
  return `path.resolve(__dirname, ${JSON.stringify(
    path.relative(outputDir, cwd),
  )})`
}

function flatten(array): any[] {
  return Array.prototype.concat.apply([], array)
}

export function flatMap<T, U>(
  array: T[],
  callbackFn: (value: T, index: number, array: T[]) => U[],
  thisArg?: any,
): U[] {
  return flatten(array.map(callbackFn, thisArg))
}
