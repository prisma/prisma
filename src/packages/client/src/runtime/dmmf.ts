import { DMMF } from '@prisma/generator-helper'
import { Dictionary, keyBy, ScalarTypeTable, keyBy2 } from './utils/common'

export class DMMFClass implements DMMF.Document {
  public datamodel: DMMF.Datamodel
  public schema: DMMF.Schema
  public mappings: DMMF.Mapping[]
  public queryType: DMMF.OutputType
  public mutationType: DMMF.OutputType
  public outputTypes: DMMF.OutputType[]
  public outputTypeMap: Dictionary<DMMF.OutputType> = {}
  public inputTypes: DMMF.InputType[]
  public inputTypeMap: Dictionary<DMMF.InputType>
  public enumMap: Dictionary<DMMF.SchemaEnum>
  public modelMap: Dictionary<DMMF.Model>
  public mappingsMap: Dictionary<DMMF.Mapping>
  public rootFieldMap: Dictionary<DMMF.SchemaField>
  constructor({ datamodel, schema, mappings }: DMMF.Document) {
    this.datamodel = datamodel
    this.schema = schema
    this.mappings = mappings
    this.enumMap = this.getEnumMap()
    this.queryType = this.getQueryType()
    this.mutationType = this.getMutationType()
    this.modelMap = this.getModelMap()

    this.outputTypes = this.getOutputTypes()

    this.outputTypeMap = this.getMergedOutputTypeMap()

    this.resolveOutputTypes(this.outputTypes)

    this.inputTypes = this.schema.inputTypes
    this.inputTypeMap = this.getInputTypeMap()
    this.resolveInputTypes(this.inputTypes)
    this.resolveFieldArgumentTypes(this.outputTypes, this.inputTypeMap)
    this.mappingsMap = this.getMappingsMap()

    // needed as references are not kept
    this.queryType = this.outputTypeMap.Query
    this.mutationType = this.outputTypeMap.Mutation
    this.outputTypes = this.outputTypes
    this.rootFieldMap = this.getRootFieldMap()
  }
  protected outputTypeToMergedOutputType = (
    outputType: DMMF.OutputType,
  ): DMMF.OutputType => {
    const model = this.modelMap[outputType.name]
    return {
      ...outputType,
      isEmbedded: model ? model.isEmbedded : false,
      fields: outputType.fields,
    }
  }
  protected resolveOutputTypes(types: DMMF.OutputType[]) {
    for (const type of types) {
      for (const field of type.fields) {
        if (
          typeof field.outputType.type === 'string' &&
          !ScalarTypeTable[field.outputType.type]
        ) {
          field.outputType.type =
            this.outputTypeMap[field.outputType.type] ||
            this.enumMap[field.outputType.type] ||
            field.outputType.type
        }
      }
      type.fieldMap = keyBy(type.fields, 'name')
    }
  }
  protected resolveInputTypes(types: DMMF.InputType[]) {
    for (const type of types) {
      for (const field of type.fields) {
        const first = field.inputTypes[0].type
        if (
          typeof first === 'string' &&
          !ScalarTypeTable[first] &&
          (this.inputTypeMap[first] || this.enumMap[first])
        ) {
          field.inputTypes[0].type =
            this.inputTypeMap[first] ||
            this.enumMap[first] ||
            field.inputTypes[0].type
        }
        const second = field.inputTypes[1] && field.inputTypes[1].type
        if (
          typeof second === 'string' &&
          !ScalarTypeTable[second] &&
          (this.inputTypeMap[second] || this.enumMap[second])
        ) {
          field.inputTypes[1].type =
            this.inputTypeMap[second] ||
            this.enumMap[second] ||
            field.inputTypes[1].type
        }
      }
      type.fieldMap = keyBy(type.fields, 'name')
    }
  }
  protected resolveFieldArgumentTypes(
    types: DMMF.OutputType[],
    inputTypeMap: Dictionary<DMMF.InputType>,
  ) {
    for (const type of types) {
      for (const field of type.fields) {
        for (const arg of field.args) {
          const first = arg.inputTypes[0].type
          if (typeof first === 'string' && !ScalarTypeTable[first]) {
            arg.inputTypes[0].type =
              inputTypeMap[first] ||
              this.enumMap[first] ||
              arg.inputTypes[0].type
          }
          const second = arg.inputTypes[1] && arg.inputTypes[1].type
          if (
            second &&
            typeof second === 'string' &&
            !ScalarTypeTable[second]
          ) {
            arg.inputTypes[1].type =
              inputTypeMap[second] ||
              this.enumMap[second] ||
              arg.inputTypes[1].type
          }
        }
      }
    }
  }
  protected getQueryType(): DMMF.OutputType {
    return this.schema.outputTypes.find((t) => t.name === 'Query')!
  }
  protected getMutationType(): DMMF.OutputType {
    return this.schema.outputTypes.find((t) => t.name === 'Mutation')!
  }
  protected getOutputTypes(): DMMF.OutputType[] {
    return this.schema.outputTypes.map(this.outputTypeToMergedOutputType)
  }
  protected getEnumMap(): Dictionary<DMMF.SchemaEnum> {
    return keyBy(this.schema.enums, 'name')
  }
  protected getModelMap(): Dictionary<DMMF.Model> {
    return keyBy(this.datamodel.models, 'name')
  }
  protected getMergedOutputTypeMap(): Dictionary<DMMF.OutputType> {
    return keyBy(this.outputTypes, 'name')
  }
  protected getInputTypeMap(): Dictionary<DMMF.InputType> {
    return keyBy(this.schema.inputTypes, 'name')
  }
  protected getMappingsMap(): Dictionary<DMMF.Mapping> {
    return keyBy(this.mappings, 'model')
  }
  protected getRootFieldMap(): Dictionary<DMMF.SchemaField> {
    return keyBy2(this.queryType.fields, this.mutationType.fields, 'name')
  }
}
