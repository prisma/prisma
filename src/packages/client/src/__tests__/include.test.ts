import stripAnsi from 'strip-ansi'
import { blog } from '../fixtures/blog'
import { DMMFClass } from '../runtime/dmmf'
import { makeDocument } from '../runtime/query'
import { getDMMF } from '../generation/getDMMF'

let dmmf
beforeAll(async () => {
  const dmmfDocument = await getDMMF({ datamodel: blog })
  dmmf = new DMMFClass(dmmfDocument)
})

describe('include validation', () => {
  test('deep include query', () => {
    const ast = {
      include: {
        author: {
          include: {
            posts: true,
          },
        },
      },
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'findManyPost',
    })

    expect(String(document)).toMatchSnapshot()
    expect(() => document.validate(ast)).not.toThrow()
  })

  test('dont allow empty include statements', () => {
    const ast = {
      include: {},
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'findManyPost',
    })

    expect(String(document)).toMatchSnapshot()
    try {
      document.validate(ast, false)
    } catch (e) {
      expect(stripAnsi(e.message)).toMatchInlineSnapshot(`
        "
        Invalid \`prisma.findManyPost()\` invocation:

        {
          include: {
        ?   author?: true,
        ?   categories?: true
          }
        }


        The \`include\` statement for type Post must not be empty. Available options are listed in green.
        "
      `)
    }
  })

  test('allow normal findMany without include for empty model', () => {
    const ast = {}

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'findManyNoRelations',
    })

    expect(String(document)).toMatchSnapshot()
    document.validate(ast, false)
  })

  test('enforce no include, if no relation', () => {
    const ast = {
      include: {},
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'findManyNoRelations',
    })

    expect(String(document)).toMatchSnapshot()
    try {
      document.validate(ast, false)
    } catch (e) {
      expect(stripAnsi(e.message)).toMatchInlineSnapshot(`
        "
        Invalid \`prisma.findManyNoRelations()\` invocation:

        {
          include: {}
        }


        NoRelations does not have any relation and therefore can't have an \`include\` statement.
        "
      `)
    }
  })

  test('enforce empty include, if no relation', () => {
    const ast = {
      include: {
        asd: true,
      },
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'findManyNoRelations',
    })

    expect(String(document)).toMatchSnapshot()
    try {
      document.validate(ast, false)
    } catch (e) {
      expect(stripAnsi(e.message)).toMatchInlineSnapshot(`
        "
        Invalid \`prisma.asd()\` invocation:

        {
          include: {
            asd: true
            ~~~
          }
        }


        Unknown field \`asd\` for include statement on model NoRelations.
        This model has no relations, so you can't use include with it.
        "
      `)
    }
  })

  // Why do we allow it with include but not with select?
  // Very simple, because with select a statement with only false properties is useless
  // Why do we throw for an empty object? Also very simple: We want to help people to explore
  // the API through errors!
  test('allow include statement with only false properties', () => {
    const ast = {
      include: {
        author: true,
      },
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'findManyPost',
    })

    expect(String(document)).toMatchSnapshot()
    expect(() => document.validate(ast)).not.toThrow()
  })

  test('allow deep include with empty object', () => {
    const ast = {
      include: {
        posts: {},
      },
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'findManyUser',
    })

    expect(String(document)).toMatchInlineSnapshot(`
      "query {
        findManyUser {
          id
          email
          name
          json
          countFloat
          countInt1
          countInt2
          countInt3
          countInt4
          countInt5
          countInt6
          lastLoginAt
          coinflips
          posts {
            id
            createdAt
            updatedAt
            published
            title
            content
            authorId
            optionnal
          }
        }
      }"
    `)
    expect(() => document.validate(ast)).not.toThrow()
  })

  test('allow deep include without another include', () => {
    const ast = {
      include: {
        posts: { take: 20 },
      },
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'findManyUser',
    })

    expect(String(document)).toMatchSnapshot()
    expect(() => document.validate(ast)).not.toThrow()
  })

  test('handle scalar fields special', () => {
    const ast = {
      include: {
        id: true,
      },
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'findManyPost',
    })

    expect(String(document)).toMatchSnapshot()
    try {
      document.validate(ast, false)
    } catch (e) {
      expect(stripAnsi(e.message)).toMatchInlineSnapshot(`
        "
        Invalid \`prisma.id()\` invocation:

        {
          include: {
            id: true,
            ~~
        ?   author?: true,
        ?   categories?: true
          }
        }


        Invalid scalar field \`id\` for include statement on model Post. Available options are listed in green.
        Note, that include statements only accept relation fields.
        "
      `)
    }
  })

  test('catch unknown field name', () => {
    const ast = {
      include: {
        mauthor: true,
      },
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'findManyPost',
    })

    expect(String(document)).toMatchSnapshot()
    try {
      document.validate(ast, false)
    } catch (e) {
      expect(stripAnsi(e.message)).toMatchInlineSnapshot(`
        "
        Invalid \`prisma.mauthor()\` invocation:

        {
          include: {
            mauthor: true,
            ~~~~~~~
        ?   author?: true,
        ?   categories?: true
          }
        }


        Unknown field \`mauthor\` for include statement on model Post. Available options are listed in green. Did you mean \`author\`?
        "
      `)
    }
  })

  test('allow include with a select', () => {
    const ast = {
      include: {
        posts: {
          take: 20,
          select: {
            id: true,
          },
        },
      },
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'findManyUser',
    })

    expect(String(document)).toMatchInlineSnapshot(`
      "query {
        findManyUser {
          id
          email
          name
          json
          countFloat
          countInt1
          countInt2
          countInt3
          countInt4
          countInt5
          countInt6
          lastLoginAt
          coinflips
          posts(take: 20) {
            id
          }
        }
      }"
    `)
    expect(() => document.validate(ast)).not.toThrow()
  })

  test('allow include with a select with an include', () => {
    const ast = {
      include: {
        posts: {
          take: 20,
          select: {
            id: true,
            author: {
              include: { posts: true },
            },
          },
        },
      },
    }

    const document = makeDocument({
      dmmf,
      select: ast,
      rootTypeName: 'query',
      rootField: 'findManyUser',
    })

    expect(String(document)).toMatchInlineSnapshot(`
      "query {
        findManyUser {
          id
          email
          name
          json
          countFloat
          countInt1
          countInt2
          countInt3
          countInt4
          countInt5
          countInt6
          lastLoginAt
          coinflips
          posts(take: 20) {
            id
            author {
              id
              email
              name
              json
              countFloat
              countInt1
              countInt2
              countInt3
              countInt4
              countInt5
              countInt6
              lastLoginAt
              coinflips
              posts {
                id
                createdAt
                updatedAt
                published
                title
                content
                authorId
                optionnal
              }
            }
          }
        }
      }"
    `)
    expect(() => document.validate(ast)).not.toThrow()
  })
})
