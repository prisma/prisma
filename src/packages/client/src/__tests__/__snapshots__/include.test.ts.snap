// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`include validation allow deep include without another include 1`] = `
"query {
  findManyUser {
    id
    email
    name
    json
    countFloat
    countInt1
    countInt2
    countInt3
    countInt4
    countInt5
    countInt6
    lastLoginAt
    coinflips
    posts(take: 20) {
      id
      createdAt
      updatedAt
      published
      title
      content
      authorId
      optionnal
    }
  }
}"
`;

exports[`include validation allow include statement with only false properties 1`] = `
"query {
  findManyPost {
    id
    createdAt
    updatedAt
    published
    title
    content
    authorId
    optionnal
    author {
      id
      email
      name
      json
      countFloat
      countInt1
      countInt2
      countInt3
      countInt4
      countInt5
      countInt6
      lastLoginAt
      coinflips
    }
  }
}"
`;

exports[`include validation allow normal findMany without include for empty model 1`] = `
"query {
  findManyNoRelations {
    id
    name
  }
}"
`;

exports[`include validation catch unknown field name 1`] = `
"query {
  mauthor {
    mauthor # INVALID_FIELD
  }
}"
`;

exports[`include validation deep include query 1`] = `
"query {
  findManyPost {
    id
    createdAt
    updatedAt
    published
    title
    content
    authorId
    optionnal
    author {
      id
      email
      name
      json
      countFloat
      countInt1
      countInt2
      countInt3
      countInt4
      countInt5
      countInt6
      lastLoginAt
      coinflips
      posts {
        id
        createdAt
        updatedAt
        published
        title
        content
        authorId
        optionnal
      }
    }
  }
}"
`;

exports[`include validation dont allow empty include statements 1`] = `
"query {
  findManyPost {
    include # INVALID_FIELD
  }
}"
`;

exports[`include validation enforce empty include, if no relation 1`] = `
"query {
  asd {
    asd # INVALID_FIELD
  }
}"
`;

exports[`include validation enforce no include, if no relation 1`] = `
"query {
  findManyNoRelations {
    include # INVALID_FIELD
  }
}"
`;

exports[`include validation handle scalar fields special 1`] = `
"query {
  id {
    id # INVALID_FIELD
  }
}"
`;
